### ISSUE TITLE
Polygon does not close, and 'equals' function does not work

### ISSUE THREAD

**Issue by YairSchwartz** [Toblerity/Shapely#410]
When pushing 4 vertices in the the polygon function, though the output should be a tuple with 5 elements, It seems that the order of the input vertices passed to the function has impact the result: sometime the polygon is described with 5 vertices and sometimes with 4 - meaning, it`s not a closed polygon (or in other words - it is not a polygon at all) It must be some bug.

Example:

```
from shapely.geometry import Polygon

print ('poly1 = ', Polygon([(620, 420, 500), (620, 420, 0), (620, 40, 0),(620, 40, 500)]))
print ('poly2 = ',Polygon([(620, 40, 500), (620, 420, 500), (620, 420, 0), (620, 40, 0)]))

>>poly1 =  POLYGON Z ((620 420 500, 620 420 0, 620 40 0, 620 40 500, 620 420 500))
>>poly2 =  POLYGON Z ((620 40 500, 620 420 500, 620 420 0, 620 40 0))
```

Also, when using .equals function - it doesn`t always work:

```
poly1 =  Polygon(([220.0, 400, 500], [220.0, 20, 500], [220.0, 20, 0], [220.0, 400, 0], [220.0, 400, 500]))
poly2 =  Polygon(([220.0, 20, 500], [220.0, 400, 500], [220.0, 400, 0], [220.0, 20, 0], [220.0, 20, 500]))

print (poly1.equals(poly2))
```

gives a False

**Comment by sgillies** [comment]
@YairSchwartz Something I haven't explained very well about Shapely is that it is a 2D geometry library where Z values are ignored by most of the constructive and analytic algorithms. Your polygons aren't strictly valid in 2D: `(620, 420, 500)` and `(620, 420, 0)` are considered to be the same point and so your Polygons don't have the required number of vertices. They each have only 2 unique vertices.

```python
>>> Polygon([(620, 420, 500), (620, 420, 0), (620, 40, 0),(620, 40, 500)]).is_valid
Too few points in geometry component at or near point 620 420 500
False
```

I think it's clear Shapely is not doing the right thing here and should raise an error instead of leaving you with perplexing and invalid polygons.

**Comment by sgillies** [comment]
We'll get this done for 1.6.3. Plan: make a list of coordinates with unique x, y values and check the length of *that*, raising a ValueError if it isn't 3 for linear rings and polygons and if it isn't 2 for line strings.

**Comment by sgillies** [comment]
After some review of the code, this is a can of worms. Here's the deal:

Shapely currently allows you to work in "anything goes" mode. You can easily create invalid or pathological geometries like `LineString([(0, 0), (0, 0)])` and get no exceptions or warnings. If you want to work in a more strict mode, you can check the validity of objects (`geom.is_valid` or `shapely.validation.explain_validity(geom)`). Our reasoning is that checking validity is an extra expense for programmers who are certain to have valid geometries.

I'm beginning to realize that checking for the xy-uniqueness of coordinates that have Z values can't fully guard against invalidity unless we check every coordinate tuple, and even then doesn't guard against accidental self-intersections. It's not an adequate substitute for calling `is_valid` post-creation. I think more documentation will be the solution to this issue.

**Comment by sgillies** [comment]
Resolved by ae435c4.
